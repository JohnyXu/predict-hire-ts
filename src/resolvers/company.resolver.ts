import Company, { ICompany } from '../models/Company';

export default {
  Query: {
    getCompanies: async (): Promise<Array<ICompany>> => {
      try {
        const companies = await Company.find();
        return companies;
      } catch (error) {
        throw new Error(error);
      }
    },
  },
};
