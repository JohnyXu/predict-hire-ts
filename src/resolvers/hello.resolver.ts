import { IResolvers } from 'graphql-tools';

const helloResolver: IResolvers = {
  Query: {
    helloWorld: (): string => {
      return 'hello world!';
    },
  },
};

export default helloResolver;
