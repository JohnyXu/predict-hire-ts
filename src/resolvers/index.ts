import helloResolver from './hello.resolver';
import companyResolver from './company.resolver';
import userResolver from './user.resolver';

export default [helloResolver, companyResolver, userResolver];
