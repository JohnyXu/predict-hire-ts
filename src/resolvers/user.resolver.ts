import { IUser } from './../models/User';
import { AuthenticationError, UserInputError } from 'apollo-server-express';
import { generateUserToken } from '../helper/JwtService';
import User from '../models/User';
import { UserEntity } from '../interface/user.types';

export default {
  Query: {
    // eslint-disable-next-line
    getUsers: async (parent, args, context): Promise<Array<IUser>> => {
      if (!context.user) {
        throw new AuthenticationError(
          'Login first and attach token with request',
        );
      }
      try {
        return await User.find();
      } catch (error) {
        throw new Error(error);
      }
    },

    // eslint-disable-next-line
    getUser: async (parent, { id }, context): Promise<IUser> => {
      if (!context.user) {
        throw new AuthenticationError(
          'Login first and attach token with request',
        );
      }

      if (!id.trim()) {
        throw new UserInputError('Empty user id');
      }

      try {
        const user = await User.findById(id);
        if (user) {
          return user;
        }
        throw new Error('User not found');
      } catch (error) {
        throw new Error(error);
      }
    },
  },

  Mutation: {
    login: async (
      // eslint-disable-next-line
      parent,
      // eslint-disable-next-line
      { loginInput: { username, password } },
    ): Promise<UserEntity> => {
      if (!username.trim()) {
        throw new UserInputError("Error: username can't be empty");
      }
      if (!password.trim()) {
        throw new UserInputError("Error: password can't be empty");
      }

      const user = await User.findOne({ username });
      if (!user) {
        throw new UserInputError('Invalid user account');
      }
      const match = await user.matchPassword(password);
      if (!match) {
        throw new UserInputError('Invalid user account');
      }
      const token: string = generateUserToken(user);
      const retUser = {
        id: user._id,
        name: user.name,
        username: user.username,
        role: user.role,
        companyId: user.companyId,
        token,
      };
      return retUser;
    },
  },
};
