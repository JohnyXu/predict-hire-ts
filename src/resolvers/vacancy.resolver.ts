import { VacancyEntity } from './../interface/vacancy.types';
import { UserInputError } from 'apollo-server-express';
import { isAdminUser, isAuthUser } from '../middleware/protectRoute';
import {
  validCreateVacancy,
  validUpdateVacancy,
} from '../validator/vacancy.validator';
import Vacancy, { IVacancy } from '../models/Vacancy';
import {
  VacancyInputEntity,
  VacancyUpdateError,
} from '../interface/vacancy.types';
import { convertExpiredAt, DateExpired, formatDate } from '../helper/date';
import moment from 'moment';

function convertExpiredAt2Date(expiredAt: string): string {
  const dateExpired: DateExpired = convertExpiredAt(expiredAt);
  if (dateExpired.count <= 0) {
    throw new UserInputError('Invalid expiredAt format');
  }
  const newExpired: string = formatDate(
    moment().add(dateExpired.count, dateExpired.unit),
  );
  return newExpired;
}

function convertIVacancy2Entity(vacany: IVacancy): VacancyEntity {
  const retVacancy: VacancyEntity = {
    id: vacany.id,
    title: vacany.title,
    description: vacany.description,
    userId: vacany.userId,
    companyId: vacany.companyId,
    expiredAt: formatDate(moment(vacany.expiredAt)),
  };
  return retVacancy;
}

export default {
  Query: {
    getVacancies: isAuthUser(
      async (): Promise<Array<VacancyEntity>> => {
        try {
          const vacancies = await Vacancy.find();
          return vacancies.map(
            (vacancy: IVacancy): VacancyEntity => {
              return convertIVacancy2Entity(vacancy);
            },
          );
        } catch (error) {
          throw new Error(error);
        }
      },
    ),

    getVacancy: isAuthUser(
      async (parent, { id }): Promise<VacancyEntity> => {
        if (!id.trim()) {
          throw new UserInputError('Vacancy id empty');
        }

        try {
          const vacancy = await Vacancy.findById(id);
          if (vacancy) {
            return convertIVacancy2Entity(vacancy);
          }
          throw new Error('Vacancy not found');
        } catch (error) {
          throw new Error(error);
        }
      },
    ),
  },

  Mutation: {
    createVacancy: isAdminUser(
      async (
        parent,
        { vacancy: { title, description, expiredAt } },
        { user },
      ) => {
        const errors: VacancyInputEntity = validCreateVacancy(
          title,
          description,
          expiredAt,
        );
        if (Object.values(errors).length > 0) {
          throw new UserInputError('Error field:', { errors });
        }

        try {
          const newExpired: string = convertExpiredAt2Date(expiredAt);
          const newVacancy = new Vacancy({
            title,
            description,
            expiredAt: newExpired,
            userId: user.id,
            companyId: user.companyId,
          });
          await newVacancy.save();
          return convertIVacancy2Entity(newVacancy);
        } catch (error) {
          throw new Error(error);
        }
      },
    ),
    updateVacancy: isAdminUser(
      async (parent, { id, vacancy: { title, description, expiredAt } }) => {
        const errors: VacancyUpdateError = validUpdateVacancy(
          id,
          title,
          description,
          expiredAt,
        );
        if (Object.values(errors).length > 0) {
          throw new UserInputError('Error field:', { errors });
        }

        try {
          const updateVacancy = await Vacancy.findById(id);
          if (!updateVacancy) {
            throw new UserInputError('Invalid vacancy id');
          }
          if (title) {
            updateVacancy.title = title;
          }
          if (description) {
            updateVacancy.description = description;
          }
          if (expiredAt) {
            updateVacancy.expiredAt = convertExpiredAt2Date(expiredAt);
          }
          await updateVacancy.save();
          return convertIVacancy2Entity(updateVacancy);
        } catch (error) {
          throw new Error(error);
        }
      },
    ),
    deleteVacancy: isAdminUser(async (parent, { id }) => {
      if (!id.trim()) {
        throw new UserInputError('Vacancy id empty');
      }

      try {
        const deleteVacancy = await Vacancy.findById(id);
        if (!deleteVacancy) {
          throw new UserInputError('vacancy not found');
        }
        await deleteVacancy.remove();
        return true;
      } catch (error) {
        throw new Error(error);
      }
    }),
  },
};
