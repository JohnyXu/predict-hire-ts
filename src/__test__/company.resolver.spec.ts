import dotenv from 'dotenv';
dotenv.config();

import { ApolloServer, gql } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';
import resolvers from '../resolvers';
import typeDefs from '../typedefs';

import connectMongo from '../helper/connect';

const server = new ApolloServer({
  typeDefs,
  resolvers,
});

const { query } = createTestClient(server);

describe('Company resolver', () => {
  beforeAll(async () => {
    await connectMongo();
  });

  test('get companies', async () => {
    const GET_COMPANIES = gql`
      query {
        getCompanies {
          id
          name
          address
        }
      }
    `;

    const {
      data: { getCompanies },
    } = await query({ query: GET_COMPANIES });
    expect(getCompanies).toEqual([
      {
        id: '5e5df7fc6953acd3dc50fe8f',
        name: 'PredictiveHire',
        address: '15 Newton St',
      },
    ]);
    expect(getCompanies).toMatchSnapshot();
  });
});
