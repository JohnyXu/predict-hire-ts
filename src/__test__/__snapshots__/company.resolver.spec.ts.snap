// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company resolver get companies 1`] = `
Array [
  Object {
    "address": "15 Newton St",
    "id": "5e5df7fc6953acd3dc50fe8f",
    "name": "PredictiveHire",
  },
]
`;
